### PR TITLE
Add support for Certificate of Trust

### PR DESCRIPTION
**Pull request by gregarndt**

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -242,6 +243,7 @@ test:
   dockerVolume: '/tmp'
 
   dockerWorkerPrivateKey: '/worker/test/docker-worker-priv.pem'
+  signingKeyLocation: '/worker/test/fixtures/gpg_signing_key.key'
```
Comment (walac) [inline config.yml]: I wonder why we need to specify this configuration twice in the file.

File: deploy/packer/app.json [hunk 2]
```diff
@@ -4,8 +4,8 @@
   "variables": {
     "npmPackage":          "",
     "templateContents":    "",
-    "hvmSourceAMI":        "ami-0710db67",
-    "pvSourceAMI":         "ami-ee17dc8e",
+    "hvmSourceAMI":        "ami-7c22f71c",
+    "pvSourceAMI":         "ami-e422f784",
```
Comment (walac) [inline deploy/packer/app.json]: Does this incorporate the recent security fix?

File: lib/features/certificate_of_trust.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
+/**
+The bulk logger writes the task stream directly to disk then uploads that file
+to s3 after the task has completed running.
+*/
```
Comment (walac) [inline lib/features/certificate_of_trust.js]: I think this comment doesn't belong to this module.

File: lib/features/certificate_of_trust.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    this.file = new temporary.File();
+
+    // Pipe the task stream to a temp file on disk.
+    this.stream = fs.createWriteStream(this.file.path);
+    task.stream.on('data', (d) => {
+      this.hash.update(d);
+    });
+
+    let gzip = zlib.createGzip();
+    task.stream.pipe(gzip).pipe(this.stream);
```
Comment (walac) [inline lib/features/certificate_of_trust.js]: A comment on what this does would be good.

File: lib/features/certificate_of_trust.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      await uploadToS3(task.queue, task.status.taskId, task.runId,
+                       logStream, 'public/logs/certified.log', expiration, {
+        'content-type': 'text/plain',
+        'content-length': stat.size,
+        'content-encoding': 'gzip'
+      });
+    } catch (err) {
+      debug(err);
+      // Unlink the temp file.
+      await fs.unlink(this.file.path);
```
Comment (walac) [inline lib/features/certificate_of_trust.js]: You can safely `unlink` the file before upload to S3. The stream keeps a file descriptor open, and the file won't be removed before it is closed.

File: lib/features/certificate_of_trust.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      workerGroup: task.runtime.workerGroup,
+      workerId: task.runtime.workerId,
+      extra: {
+        publicIpAddress: task.runtime.publicIp,
+        privateIpAddress: task.runtime.privateIp,
+        imageHash: task.imageHash,
+      }
+    };
+
+    ['instanceId', 'instanceType', 'region'].forEach((tag) => {
```
Comment (walac) [inline lib/features/certificate_of_trust.js]: nit: the parenthesis around tag are unnecessary.

**Comment by petemoore** [comment]
@escapewindow feel free to provide feedback here too

**Comment by escapewindow** [comment]
@petemoore Thanks for the heads up!  I verified the output in bug 1284991.  I'm not a js expert, so I'll leave the code review to those who are.

File: config.yml (hunk 1 above)
Comment (gregarndt) [inline config.yml]: they are different locations.  One as the default, one for tests.

File: deploy/packer/app.json (hunk 2 above)
Comment (gregarndt) [inline deploy/packer/app.json]: It should?  I pulled down master and rebuilt the base

Comment (gregarndt) [inline deploy/packer/app.json]: yup, this was in the base.json in my cot branch:
    {
      "type": "shell",
      "inline": [
        "sudo bash -c 'echo net.ipv4.tcp_challenge_ack_limit = 999999999 >> /etc/sysctl.conf'"
      ]
    }

File: lib/features/certificate_of_trust.js (hunk 3 above)
Comment (gregarndt) [inline lib/features/certificate_of_trust.js]: quite right, I've updated it.  Thanks for catching that.

**Comment by gregarndt** [comment]
@walac I think this addressed your comments now.

**Comment by petemoore** [comment]
@gregarndt any idea why http://schemas.taskcluster.net/docker-worker/v1/payload.json didn't get updated when this merged?

File: lib/features.js [hunk 7]
```diff
@@ -16,6 +16,15 @@ const features = {
     module: require('./features/local_live_log')
   },
 
+  generateCertificate: {
+    title: 'Enable generation of a certificate for Chain of Trust',
+    description: 'A certificate should be generated which will include information ' +
+                 'for downstream tasks to build a level of trust for the artifacts' +
+                 'produced by the task and the environment it ran in.',
```
Comment (petemoore) [inline lib/features.js]: missing space between `artifacts` and `produced`

File: test/integration/certificate_of_trust.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    let key = openpgp.key.readArmored(armoredKey);
+    let opts = {
+      publicKeys: key.keys,
+      message: openpgp.cleartext.readArmored(signedCertificate)
+    };
+    let verified = await openpgp.verify(opts);
+
+    assert(verified.signatures[0].valid, 'Certificate does not appear to be valid');
+
+    // computer the hash of the live_backing.log which should be the same as the
```
Comment (petemoore) [inline test/integration/certificate_of_trust.js]: "compute"

**Pull request merged**